File: python/geometry/polygon/polygon_clip.py
#!/usr/bin/python
#\file    polygon_clip.py
#\brief   Test Sutherland-Hodgman-Algorithm
#\author  Akihiko Yamaguchi, info@akihikoy.net
#\version 0.1
#\date    Jul.01, 2016


import math,random

#Clip subject_polygon by clip_polygon with Sutherland-Hodgman-Algorithm.
#subject_polygon is an arbitrary polygon, clip_polygon is a convex polygon.
#Output polygon is counterclockwise.
#Ref: https://rosettacode.org/wiki/Sutherland-Hodgman_polygon_clipping
def ClipPolygon(subject_polygon, clip_polygon):
  def is_left_of(edge_1, edge_2, test):
    tmp1 = [edge_2[0] - edge_1[0], edge_2[1] - edge_1[1]]
    tmp2 = [test[0] - edge_2[0], test[1] - edge_2[1]]
    x = (tmp1[0] * tmp2[1]) - (tmp1[1] * tmp2[0])
    if x < 0:  return False
    elif x > 0:  return True
    else:  return None  # Colinear points;

  def is_clockwise(polygon):
    for p in polygon:
      is_left = is_left_of(polygon[0], polygon[1], p);
      if is_left != None:  #some of the points may be colinear.  That's ok as long as the overall is a polygon
        return not is_left
    return None #All the points in the polygon are colinear

  def inside(p):
    return(cp2[0]-cp1[0])*(p[1]-cp1[1]) > (cp2[1]-cp1[1])*(p[0]-cp1[0])

  def compute_intersection():
    dc = [ cp1[0] - cp2[0], cp1[1] - cp2[1] ]
    dp = [ s[0] - e[0], s[1] - e[1] ]
    n1 = cp1[0] * cp2[1] - cp1[1] * cp2[0]
    n2 = s[0] * e[1] - s[1] * e[0]
    n3 = 1.0 / (dc[0] * dp[1] - dc[1] * dp[0])
    return [(n1*dp[0] - n2*dc[0]) * n3, (n1*dp[1] - n2*dc[1]) * n3]

  ic = is_clockwise(subject_polygon)
  if ic is None:  return []
  if ic:  subject_polygon.reverse()
  ic = is_clockwise(clip_polygon)
  #print 'is_clockwise(clip_polygon)=',ic
  if ic is None:  return []
  if ic:  clip_polygon.reverse()

  output_list = subject_polygon
  cp1 = clip_polygon[-1]
  #print 'clip_polygon=',clip_polygon

  for clip_vertex in clip_polygon:
    cp2 = clip_vertex
    input_list = output_list
    output_list = []
    #print 'input_list =',input_list
    if len(input_list)==0:
      return []
    s = input_list[-1]
    #print '  s=',s
    #print '  cp1=',cp1
    #print '  cp2=',cp2

    for subject_vertex in input_list:
      e = subject_vertex
      #print '  e=',e,inside(e),inside(s)
      if inside(e):
        if not inside(s):
          output_list.append(compute_intersection())
        output_list.append(e)
      elif inside(s):
        output_list.append(compute_intersection())
      s = e
    cp1 = cp2
  return(output_list)

#Float version of range
def FRange1(x1,x2,num_div):
  return [x1+(x2-x1)*x/float(num_div) for x in range(num_div+1)]

#Generate a random number of uniform distribution of specified bound.
def Rand(xmin=-0.5,xmax=0.5):
  return random.random()*(xmax-xmin)+xmin

def Main():
  #Generate polygon:
  #polygon= [[50, 150], [200, 50], [350, 150], [350, 300],
            #[250, 300], [200, 250], [150, 350], [100, 250],
            #[100, 200]]
  polygon= []
  N= 100
  for i in xrange(N):
    th= (float(i)+Rand(-0.4,0.4))*2.0*math.pi/float(N)
    r= Rand(1.0, 100.0)
    x= r*math.cos(th)
    y= r*math.sin(th)
    polygon.append([x,y])

  #rect= [[30.0,30.0],[30.0,-30.0],[-30.0,-30.0],[-30.0,30.0]]
  #rect= [[10*(p[0]+50),10*(p[1]+50)] for p in rect]
  #rect= [[30.0,30.0],[-30.0,30.0],[-30.0,-30.0],[30.0,-30.0]]
  #rect= [[100, 100], [300, 100], [300, 300], [100, 300]]
  #p1x= Rand(1.0, 100.0); p1y= Rand(1.0, 100.0); p2x= Rand(1.0, 100.0); p2y= Rand(1.0, 100.0)
  p1x= Rand(-100.0, 100.0); p1y= Rand(-100, 100.0); p2x= Rand(-100, 100.0); p2y= Rand(-100, 100.0)
  rect= [[p1x,p1y],[p2x,p1y],[p2x,p2y],[p1x,p2y]]
  #rect= [[p1x,p1y],[p2x,p2y],[p1x,p1y],[p2x,p2y]]

  #polygon= [[-1.69021, -87.6474],[-40.5151, -2.28697],[-25.867, 44.6599],[29.6493, -3.34294]]
  #rect= [[-88.5993, 13.5659],[57.7085, 13.5659],[57.7085, -55.5519],[-88.5993, -55.5519]]
  polygon= [[45.672222, -2.9928033],[1.6948826, 8.3420715],[-77.632523, 33.676598],[-29.458261, -45.078251]]
  rect= [[-61.265263, -16.213083],[-37.288738, -16.213083],[-37.288738, 72.973518],[-61.265263, 72.973518]]

  polygon2= ClipPolygon(polygon, rect)
  #try:
    #polygon2= ClipPolygon(polygon, rect)
  #except Exception as e:
    #print 'Error'
    #print '  e: ',e
    #print '  type: ',type(e)
    #print '  args: ',e.args
    #print '  message: ',e.message
    #polygon2= []

  def save_poly(poly,name):
    fp= open('/tmp/'+name,'w')
    for p in poly:
      fp.write('{x} {y}\n'.format(x=p[0],y=p[1]))
    if len(poly)>0:
      fp.write('{x} {y}\n'.format(x=poly[0][0],y=poly[0][1]))  #closing loop
    fp.close()
  save_poly(polygon,'polygon.dat')
  save_poly(rect,'clip_rect.dat')
  save_poly(polygon2,'polygon_clipped.dat')


def PlotGraphs():
  print 'Plotting graphs..'
  import os
  commands=[
    '''qplot -x2 aaa
      /tmp/polygon.dat w l lw 2 t '"Original"'
      /tmp/clip_rect.dat w l lw 2 t '"Rect"'
      /tmp/polygon_clipped.dat w l lw 3 t '"Clipped"'
      &''',
    '''''',
    '''''',
    ]
  for cmd in commands:
    if cmd!='':
      cmd= ' '.join(cmd.splitlines())
      print '###',cmd
      os.system(cmd)

  print '##########################'
  print '###Press enter to close###'
  print '##########################'
  raw_input()
  os.system('qplot -x2kill aaa')

if __name__=='__main__':
  import sys
  if len(sys.argv)>1 and sys.argv[1] in ('p','plot','Plot','PLOT'):
    PlotGraphs()
    sys.exit(0)
  Main()

File: opencv/cpp/polygon_clip.cpp
//-------------------------------------------------------------------------------------------
/*! \file    polygon_clip.cpp
    \brief   certain c++ source file
    \author  Akihiko Yamaguchi, info@akihikoy.net
    \version 0.1
    \date    Jul.01, 2016

g++ -g -Wall -O2 -o polygon_clip.out polygon_clip.cpp -lopencv_core -lopencv_imgproc -lopencv_highgui -lopencv_videoio

qplot -x /tmp/polygon.dat w l lw 2 t '"original"' /tmp/clip_rect.dat w l lw 2 t '"clip"' /tmp/polygon_clipped.dat w l lw 4 t '"clipped"'
*/
//-------------------------------------------------------------------------------------------
#include <opencv2/core/core.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <iostream>
//-------------------------------------------------------------------------------------------
namespace loco_rabbits
{

namespace ns_polygon_clip
{
#define TP cv::Point_<t_value>
#define ROW Row<t_value>
#define ADD Add<t_value>

template<typename t_value>
inline TP Row(const cv::Mat &points2d, int r)
{
  if(r<0)  r= points2d.rows+r;
  return TP(points2d.at<t_value>(r,0), points2d.at<t_value>(r,1));
}

template<typename t_value>
inline void Add(cv::Mat &points2d, const TP &p)
{
  cv::Mat v(p);
  v= v.t();
  points2d.push_back(v);
}

template<typename t_value>
inline int IsLeftOf(const TP &edge_1, const TP &edge_2, const TP &test)
{
  TP tmp1(edge_2.x - edge_1.x, edge_2.y - edge_1.y);
  TP tmp2(test.x - edge_2.x, test.y - edge_2.y);
  t_value x = (tmp1.x * tmp2.y) - (tmp1.y * tmp2.x);
  if(x < 0)  return 0;
  else if(x > 0)  return 1;
  else  return -1;  // Colinear points
}

template<typename t_value>
int IsClockwise(const cv::Mat &polygon)
{
  if(polygon.rows<3)  return -1;
  TP p0= ROW(polygon,0);
  TP p1= ROW(polygon,1);
  int isLeft(-1);
  for(int r(0),r_end(polygon.rows); r<r_end; ++r)
  {
    isLeft= IsLeftOf<t_value>(p0, p1, ROW(polygon,r));
    if(isLeft>=0)  // some of the points may be colinear.  That's ok as long as the overall is a polygon
      return isLeft==0 ? 1 : 0;
  }
  return -1;  // All the points in the polygon are colinear
}

template<typename t_value>
inline bool IsInside(const TP &cp1, const TP &cp2, const TP &p)
{
  return (cp2.x-cp1.x)*(p.y-cp1.y) > (cp2.y-cp1.y)*(p.x-cp1.x);
}

template<typename t_value>
inline TP ComputeIntersection(const TP &cp1, const TP &cp2, const TP &s, const TP &e)
{
  TP dc(cp1.x - cp2.x, cp1.y - cp2.y);
  TP dp(s.x - e.x, s.y - e.y);
  t_value n1= cp1.x * cp2.y - cp1.y * cp2.x;
  t_value n2= s.x * e.y - s.y * e.x;
  t_value n3= 1.0 / (dc.x * dp.y - dc.y * dp.x);
  return TP((n1*dp.x - n2*dc.x) * n3, (n1*dp.y - n2*dc.y) * n3);
}

template<typename t_value>
cv::Mat ClipPolygon_(const cv::Mat &polygon_subject_in, const cv::Mat &polygon_clip_in)
{
  cv::Mat polygon_empty(0,2,cv::DataType<t_value>::type);
  cv::Mat polygon_subject, polygon_clip;
  switch(IsClockwise<t_value>(polygon_subject_in))
  {
  case -1:
    std::cerr<<"polygon_subject: All the points are colinear"<<std::endl;
    return polygon_empty;
  case  0:  polygon_subject_in.copyTo(polygon_subject);  break;
  case +1:  cv::flip(polygon_subject_in, polygon_subject, 0);  break;
  }
//*DBG*/std::cerr<<"polygon_clip="<<polygon_clip_in<<std::endl;
//*DBG*/std::cerr<<"IsClockwise<t_value>(polygon_clip_in)="<<IsClockwise<t_value>(polygon_clip_in)<<std::endl;
  switch(IsClockwise<t_value>(polygon_clip_in))
  {
  case -1:
    std::cerr<<"polygon_clip: All the points are colinear"<<std::endl;
    return polygon_empty;
  case  0:  polygon_clip_in.copyTo(polygon_clip);  break;
  case +1:  cv::flip(polygon_clip_in, polygon_clip, 0);  break;
  }

  cv::Mat output_list= polygon_subject;
  TP cp1= ROW(polygon_clip,-1);
//*DBG*/std::cerr<<"polygon_clip="<<polygon_clip<<std::endl;

  for(int i_pc(0),i_pc_end(polygon_clip.rows); i_pc<i_pc_end; ++i_pc)
  {
    TP cp2= ROW(polygon_clip,i_pc);
    cv::Mat input_list= output_list;
    output_list= cv::Mat(0,2,cv::DataType<t_value>::type);
//*DBG*/std::cerr<<"in,out-list="<<input_list<<", "<<output_list<<std::endl;
    if(input_list.rows==0)  return polygon_empty;
    TP s= ROW(input_list,-1);
//*DBG*/std::cerr<<"  s="<<s<<std::endl;
//*DBG*/std::cerr<<"  cp1="<<cp1<<std::endl;
//*DBG*/std::cerr<<"  cp2="<<cp2<<std::endl;

    for(int i_in(0),i_in_end(input_list.rows); i_in<i_in_end; ++i_in)
    {
      TP e= ROW(input_list,i_in);
//*DBG*/std::cerr<<"  e="<<e<<", "<<IsInside<t_value>(cp1,cp2,e)<<", "<<IsInside<t_value>(cp1,cp2,s)<<std::endl;
      if(IsInside<t_value>(cp1,cp2,e))
      {
        if(!IsInside<t_value>(cp1,cp2,s))
          ADD(output_list, ComputeIntersection<t_value>(cp1, cp2, s, e));
        ADD(output_list, e);
      }
      else if(IsInside<t_value>(cp1,cp2,s))
        ADD(output_list, ComputeIntersection<t_value>(cp1, cp2, s, e));
      s= e;
    }
    cp1= cp2;
  }
  return output_list;
}
#undef TP
#undef ROW
#undef ADD
}  // ns_polygon_clip

cv::Mat ClipPolygon(const cv::Mat &polygon_subject, const cv::Mat &polygon_clip)
{
  using namespace ns_polygon_clip;
  assert(polygon_subject.type()==polygon_clip.type());
  assert(polygon_subject.cols==2);
  assert(polygon_clip.cols==2);
  switch(polygon_subject.type())
  {
  case CV_32F:  return ClipPolygon_<float>(polygon_subject, polygon_clip);
  case CV_64F:  return ClipPolygon_<double>(polygon_subject, polygon_clip);
  case CV_16S:  return ClipPolygon_<short>(polygon_subject, polygon_clip);
  case CV_32S:  return ClipPolygon_<int>(polygon_subject, polygon_clip);
  }
  throw;
}
//-------------------------------------------------------------------------------------------


}
//-------------------------------------------------------------------------------------------
#include <cmath>
#include <cstdlib>
#include <fstream>
//-------------------------------------------------------------------------------------------
using namespace std;
// using namespace boost;
using namespace loco_rabbits;
//-------------------------------------------------------------------------------------------
// #define print(var) PrintContainer((var), #var"= ")
// #define print(var) std::cout<<#var"= "<<(var)<<std::endl
//-------------------------------------------------------------------------------------------

inline unsigned Srand(void)
{
  unsigned seed ((unsigned)time(NULL));
  srand(seed);
  return seed;
}
inline double Rand (const double &max)
{
  return (max)*static_cast<double>(rand()) / static_cast<double>(RAND_MAX);
}
inline double Rand (const double &min, const double &max)
{
  return Rand(max - min) + min;
}

void SavePoly(const cv::Mat &points2d, const std::string &file_name)
{
  ofstream ofs(file_name.c_str());
  if(points2d.rows==0)  return;
  for(int r(0),r_end(points2d.rows); r<r_end; ++r)
    ofs<<points2d.at<float>(r,0)<<" "<<points2d.at<float>(r,1)<<endl;
  ofs<<points2d.at<float>(0,0)<<" "<<points2d.at<float>(0,1)<<endl;
}

int main(int argc, char**argv)
{
  using namespace ns_polygon_clip;
  Srand();
  cv::Mat polygon(0,2,CV_32F);
  int N= 100;
  for(int i(0);i<N;++i)
  {
    float th= (float(i)+Rand(-0.4,0.4))*2.0*M_PI/float(N);
    float r= Rand(1.0, 100.0);
    float x= r*cos(th);
    float y= r*sin(th);
    Add<float>(polygon, cv::Point_<float>(x,y));
  }
  float p1x,p1y, p2x,p2y;
  // p1x= Rand(1.0, 100.0); p1y= Rand(1.0, 100.0); p2x= Rand(1.0, 100.0); p2y= Rand(1.0, 100.0);
  p1x= Rand(-100.0, 100.0); p1y= Rand(-100, 100.0); p2x= Rand(-100, 100.0); p2y= Rand(-100, 100.0);
  cv::Mat rect= (cv::Mat_<float>(4,2)<<p1x,p1y, p2x,p1y, p2x,p2y, p1x,p2y);
  // std::cerr<<"polygon="<<polygon<<std::endl;
  // std::cerr<<"rect="<<rect<<std::endl;

  // cv::Mat polygon= (cv::Mat_<float>(4,2)<<-1.69021, -87.6474, -40.5151, -2.28697, -25.867, 44.6599, 29.6493, -3.34294);
  // cv::Mat rect= (cv::Mat_<float>(4,2)<<-88.5993, 13.5659, 57.7085, 13.5659, 57.7085, -55.5519, -88.5993, -55.5519);
  // cv::Mat polygon= (cv::Mat_<float>(4,2)<<-8.52197, -19.9693,-45.7384, 17.8315,24.4617, 78.9929,61.4862, 30.6966);
  // cv::Mat rect= (cv::Mat_<float>(4,2)<<-45.4064, -10.5343,2.6671, -10.5343,2.6671, -28.4372,-45.4064, -28.4372);
  // cv::Mat polygon= (cv::Mat_<float>(4,2)<<45.672222, -2.9928033,1.6948826, 8.3420715,-77.632523, 33.676598,-29.458261, -45.078251);
  // cv::Mat rect= (cv::Mat_<float>(4,2)<<-61.265263, -16.213083,-37.288738, -16.213083,-37.288738, 72.973518,-61.265263, 72.973518);

  cv::Mat polygon2= ClipPolygon(polygon, rect);

  SavePoly(polygon, "/tmp/polygon.dat");
  SavePoly(rect, "/tmp/clip_rect.dat");
  SavePoly(polygon2, "/tmp/polygon_clipped.dat");

  return 0;
}
//-------------------------------------------------------------------------------------------
